### PR TITLE
Expo update

### PR DESCRIPTION
**Pull request by lpusok**
https://bitrise.atlassian.net/browse/STEP-137

Expo project are now detected if the expo key is present in app.json (the expo/ios or expo/android keys are optional still).

Added an additional questions to ask for android package name and ios bundle ID; to help with ejecting the project. These are set in a script step in the workflow.
Added a question, wether we should run the expo CLI publish command and only then the Expo username and password is asked.

Deprecated ExpoKit as it is deprecated (https://docs.expo.io/expokit/expokit/)

If native ios/android projects are present, use the same workflow as non-epo react-native.

What is the difference between bare, managed and ExpoKit workflow? https://stackoverflow.com/a/57792159

Checked if *Expo Eject* publish step takes a long time, it does not (it seems to upload JS bundle and metadata only).

### CODE REVIEW

**Review by godrei — CHANGES_REQUESTED** [review]
Pls check my comments

File: _tests/integration/android_test.go [hunk 1]
```diff
@@ -22,6 +22,13 @@ func gitClone(t *testing.T, dir, uri string) {
 	require.NoError(t, g.Clone(uri).Run())
 }
 
+func gitCloneBranch(t *testing.T, dir, uri, branch string) {
```
[the remark this answers is not in the document]
Comment (lpusok) [reply _tests/integration/android_test.go:25]: If we suppose the default branch can be hardcoded as master, then it is possible.

File: _tests/integration/manual_config_test.go [hunk 2]
```diff
@@ -798,415 +799,507 @@ var customConfigResultYML = fmt.Sprintf(`options:
...
-                      export methods in the same build.
-                    env_key: BITRISE_EXPORT_METHOD
-                    type: selector
+                    title: The iOS project path generated by the 'expo eject' process
+                    summary: 'The relative location of the Xcode workspace, after
+                      running ''expo eject''. For example: ''./ios/myproject.xcworkspace''.
+                      Needed to eject to the bare workflow, managed workflow is not
+                      supported (https://docs.expo.io/bare/customizing/).'
+                    env_key: BITRISE_PROJECT_PATH
+                    type: user_input_optional
```
[the remark this answers is not in the document]
Comment (lpusok) [reply _tests/integration/manual_config_test.go]: No it is not, but used this option as allows to set a default value like: `./ios/< PROJECT NAME >.xcworkspace`

File: scanners/reactnative/expo.go [hunk 3]
```diff
@@ -63,204 +57,133 @@ const (
...
 	}
 
-	var moduleOption *models.OptionNode
-	if relPackageJSONDir == "" {
-		projectLocationOption := models.NewOption(android.ProjectLocationInputTitle, android.ProjectLocationInputSummary, android.ProjectLocationInputEnvKey, models.TypeSelector)
-		for _, exportMethod := range ios.IosExportMethods {
-			exportMethodOption.AddOption(exportMethod, projectLocationOption)
+	rootNode := iosNode
+	if iosNode != nil {
+		if androidNode != nil {
```
[the remark this answers is not in the document]
Comment (lpusok) [reply scanners/reactnative/expo.go:163]: It was possible originally, but not at the moment as did not find a way to determine if ios/android is used by parsing app.json.

File: _tests/integration/android_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lpusok) [reply _tests/integration/android_test.go:25]: Did not change as for example iOS App Clip sample uses main branch.

File: _tests/integration/manual_config_test.go [hunk 4]
```diff
@@ -798,415 +799,507 @@ var customConfigResultYML = fmt.Sprintf(`options:
...
-                    summary: The export method used to create an .ipa file in your
-                      builds, stored as an Environment Variable. You can change this
-                      at any time, or even create several .ipa files with different
-                      export methods in the same build.
-                    env_key: BITRISE_EXPORT_METHOD
-                    type: selector
+                    title: The iOS project path generated by the 'expo eject' process
+                    summary: 'The relative location of the Xcode workspace, after
+                      running ''expo eject''. For example: ''./ios/myproject.xcworkspace''.
+                      Needed to eject to the bare workflow, managed workflow is not
```
[the remark this answers is not in the document]
Comment (lpusok) [reply _tests/integration/manual_config_test.go]: Reworded.

File: _tests/integration/manual_config_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lpusok) [reply _tests/integration/manual_config_test.go]: Decided to remove the default value, and the optional.

File: _tests/integration/manual_config_test.go [hunk 5]
```diff
@@ -798,415 +799,507 @@ var customConfigResultYML = fmt.Sprintf(`options:
...
+                      supported (https://docs.expo.io/bare/customizing/).'
+                    env_key: BITRISE_PROJECT_PATH
+                    type: user_input_optional
                     value_map:
-                      ad-hoc:
-                        title: Project root directory
-                        summary: The directory of the `+"`app.json`"+` or `+"`package.json`"+`
-                          file of your React Native project.
-                        env_key: WORKDIR
+                      ./ios/< PROJECT NAME >.xcworkspace:
```
[the remark this answers is not in the document]
Comment (lpusok) [reply _tests/integration/manual_config_test.go]: It was valid, but only as a  default value that can be changed. Decided to remove it.

File: scanners/reactnative/expo.go [hunk 6]
```diff
@@ -63,204 +74,154 @@ const (
...
+}
+
 func appJSONError(appJSONPth, reason, explanation string) error {
 	return fmt.Errorf("app.json file (%s) %s\n%s", appJSONPth, reason, explanation)
 }
 
 // expoOptions implements ScannerInterface.Options function for Expo based React Native projects.
 func (scanner *Scanner) expoOptions() (models.OptionNode, models.Warnings, error) {
 	warnings := models.Warnings{}
+	log.TPrintf("Project name: %v", scanner.expoSettings.name)
```
[the remark this answers is not in the document]
Comment (lpusok) [reply scanners/reactnative/expo.go]: Moved after precondition check.

File: scanners/reactnative/expo.go [hunk 7]
```diff
@@ -63,204 +57,133 @@ const (
...
-		for _, exportMethod := range ios.IosExportMethods {
-			exportMethodOption.AddOption(exportMethod, workDirOption)
-		}
-
-		projectLocationOption := models.NewOption(android.ProjectLocationInputTitle, android.ProjectLocationInputSummary, android.ProjectLocationInputEnvKey, models.TypeSelector)
-		workDirOption.AddOption(relPackageJSONDir, projectLocationOption)
-
-		moduleOption = models.NewOption(android.ModuleInputTitle, android.ModuleInputSummary, android.ModuleInputEnvKey, models.TypeUserInput)
-		projectLocationOption.AddOption(filepath.Join(relPackageJSONDir, "android"), moduleOption)
+		rootNode = androidNode
```
[the remark this answers is not in the document]
Comment (lpusok) [reply scanners/reactnative/expo.go:195]: Added precondition check at the beginning of the function.

File: scanners/reactnative/reactnative.go [hunk 8]
```diff
@@ -48,24 +50,87 @@ func (Scanner) Name() string {
...
+	appJSON, err := fileutil.ReadStringFromFile(appJSONPth)
+	if err != nil {
+		return nil, err
+	}
+	var app serialized.Object
+	if err := json.Unmarshal([]byte(appJSON), &app); err != nil {
+		return nil, err
+	}
+
+	errorMessage := `To eject native projects non-interactively, the Expo project app.json file needs to contain:
```
Comment (godrei) [inline scanners/reactnative/reactnative.go:95]: these fields are not required currently

**Review by godrei — CHANGES_REQUESTED** [review]
pls see my comment

File: scanners/reactnative/expo.go [hunk 9]
```diff
@@ -63,204 +57,133 @@ const (
...
+			iosNode = models.NewOption(iosBundleIDInputTitle, iosBundleIDInputSummary, iosBundleIDEnvKey, models.TypeUserInput)
+			iosNode.AddOption("", projectPathOption)
+		} else {
+			iosNode = projectPathOption
+		}
 
-	// ensure app.json contains the required information (for non interactive eject)
-	// and predict the ejected project name
-	var projectName string
+		developmentTeamOption := models.NewOption(iosDevelopmentTeamInputTitle, iosDevelopmentTeamInputSummary, "BITRISE_IOS_DEVELOPMENT_TEAM", models.TypeUserInput)
```
[the remark this answers is not in the document]
Comment (lpusok) [reply scanners/reactnative/expo.go]: Moved

File: scanners/reactnative/expo.go [hunk 10]
```diff
@@ -63,204 +57,133 @@ const (
...
+		var projectSettingNode *models.OptionNode
+		var moduleOption *models.OptionNode
+		if relPackageJSONDir == "" {
+			projectSettingNode = models.NewOption(android.ProjectLocationInputTitle, android.ProjectLocationInputSummary, android.ProjectLocationInputEnvKey, models.TypeSelector)
 
-	log.TPrintf("Project name: %v", projectName)
+			moduleOption = models.NewOption(android.ModuleInputTitle, android.ModuleInputSummary, android.ModuleInputEnvKey, models.TypeUserInput)
+			projectSettingNode.AddOption("./android", moduleOption)
+		} else {
+			projectSettingNode = models.NewOption(projectRootDirInputTitle, projectRootDirInputSummary, "WORKDIR", models.TypeSelector)
```
[the remark this answers is not in the document]
Comment (lpusok) [reply scanners/reactnative/expo.go]: Moved

File: scanners/reactnative/reactnative.go (hunk 8 above)
Comment (lpusok) [reply scanners/reactnative/reactnative.go]: Discussed, improved logging.

**Review by godrei — APPROVED** [review]

**Pull request merged**